// react-dom/index.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, (global.index = global.index || {}, global.index.js = factory()));
})(this, (function () { 'use strict';

  const FunctionComponent = 0;
  const HostRoot = 3;
  const HostComponent = 5;
  const HostText = 6;

  const NoFlags = 0b00000001;
  const Placement = 0b00000010;
  const Update = 0b00000100;
  const ChildDeletion = 0b00001000;
  const MuationMask = Placement | Update | ChildDeletion;

  class FiberNode {
      type;
      tag;
      pendingProps;
      key;
      stateNode;
      return;
      sibling;
      child;
      index;
      ref;
      memoizeProps;
      alternate;
      flags;
      subtreeFlags;
      updateQueue;
      memoizedState;
      constructor(tag, pendingProps, key) {
          // 示例属性
          this.tag = tag;
          this.key = key;
          // HostComponent -> <div> div Dom
          this.stateNode = null;
          // FunctionComponent ->  () => {}
          this.type = null;
          // 构成树状结构
          this.return = null;
          this.sibling = null;
          this.child = null;
          this.index = 0;
          this.ref = null;
          // 作为工作单元
          this.pendingProps = pendingProps;
          this.memoizeProps = null;
          this.memoizedState = null;
          this.alternate = null;
          this.updateQueue = null;
          // 副作用
          this.flags = NoFlags;
          this.subtreeFlags = NoFlags;
      }
  }
  class FiberRootNode {
      container;
      current;
      finishedWork;
      constructor(container, hostRootFiber) {
          this.container = container;
          this.current = hostRootFiber;
          hostRootFiber.stateNode = this;
          this.finishedWork = null;
      }
  }
  const createWorkInProgress = (current, pendingProps) => {
      let wip = current.alternate;
      if (wip === null) {
          // mount
          wip = new FiberNode(current.tag, pendingProps, current.key);
          wip.stateNode = current.stateNode;
          current.alternate = wip;
      }
      else {
          wip.pendingProps = pendingProps;
          wip.flags = NoFlags;
      }
      wip.type = current.type;
      wip.updateQueue = current.updateQueue;
      wip.child = current.child;
      wip.memoizeProps = current.memoizeProps;
      wip.memoizedState = current.memoizedState;
      return wip;
  };
  function createFiberFromElement(element) {
      const { type, key, props } = element;
      let FiberTag = FunctionComponent;
      if (typeof type === "string") {
          FiberTag = HostComponent;
          // eslint-disable-next-line no-undef
      }
      else if (typeof type !== "function" && true) {
          console.warn("");
      }
      const fiber = new FiberNode(FiberTag, props, key);
      fiber.type = type;
      return fiber;
  }

  const createUpdate = (action) => {
      return {
          action,
      };
  };
  const createUpdateQueue = () => {
      return {
          shared: {
              pending: null,
          },
      };
  };
  const enqueueUpdate = (updateQueue, update) => {
      updateQueue.shared.pending = update;
  };
  const processUpdateQueue = (baseState, pendingUpdate) => {
      const result = {
          memoizedState: baseState,
      };
      if (pendingUpdate !== null) {
          const action = pendingUpdate.action;
          if (action instanceof Function) {
              result.memoizedState = action(baseState);
          }
          else {
              result.memoizedState = action;
          }
      }
      return result;
  };

  const supportSymbol = typeof Symbol === "function" && Symbol.for;
  const REACT_ELEMENT_TYPE = supportSymbol
      ? Symbol.for("react.element")
      : 0xeac7;

  function childReconciler(shouldTrackEffects) {
      return function reconilcChildFibers(returnFiber, currentFiber, newChild) {
          function reconcileSingleElement(returnFiber, currentFiber, element) {
              const fiber = createFiberFromElement(element);
              fiber.return = returnFiber;
              return fiber;
          }
          function placeSingChild(fiber) {
              if (shouldTrackEffects && fiber.alternate === null) {
                  fiber.flags |= Placement;
              }
              return fiber;
          }
          function reconcileSingleTextNode(returnFiber, currentFiber, content) {
              const fiber = new FiberNode(HostText, { content }, null);
              fiber.return = returnFiber;
              return fiber;
          }
          if (typeof newChild === "object" && newChild !== null) {
              switch (newChild.$$typeof) {
                  case REACT_ELEMENT_TYPE:
                      return placeSingChild(reconcileSingleElement(returnFiber, currentFiber, newChild));
                  default:
                      // eslint-disable-next-line no-undef
                      {
                          console.warn("");
                      }
                      break;
              }
          }
          if (typeof newChild === "string" && typeof newChild === "number") {
              return placeSingChild(reconcileSingleTextNode(returnFiber, currentFiber, newChild));
          }
          return null;
      };
  }
  const reconcileChildFiber = childReconciler(true);
  const mountChildFiber = childReconciler(false);

  // 递归中的递阶段
  const beginWork = (wip) => {
      // 比较，再返回子FilberNode
      switch (wip.tag) {
          case HostRoot:
              return updateHostRoot(wip);
          case HostComponent:
              return updateHostComponent(wip);
          case HostText:
              return null;
          default:
              // eslint-disable-next-line no-undef
              {
                  console.warn("BeginWork未实现的类型");
              }
              break;
      }
      return null;
  };
  function updateHostRoot(wip) {
      const baseState = wip.memoizedState;
      const updateQueue = wip.updateQueue;
      const pending = updateQueue.shared.pending;
      updateQueue.shared.pending = null;
      const { memoizedState } = processUpdateQueue(baseState, pending);
      wip.memoizedState = memoizedState;
      const nextChildren = wip.memoizedState;
      reconileChildren(wip, nextChildren);
      return wip.child;
  }
  function updateHostComponent(wip) {
      const nextProps = wip.pendingProps;
      const nextChildren = nextProps.children;
      reconileChildren(wip, nextChildren);
      return wip.child;
  }
  function reconileChildren(wip, children) {
      const current = wip.alternate;
      if (current !== null) {
          // update
          wip.child = reconcileChildFiber(wip, current?.child, children);
      }
      else {
          // mount
          wip.child = mountChildFiber(wip, null, children);
      }
  }

  // export const createInstance = (type: string, props: any) => {
  const createInstance = (type) => {
      // eslint-disable-next-line no-undef
      const element = document.createElement(type);
      return element;
  };
  const appendInitalChild = (parent, child) => {
      parent.appendChild(child);
  };
  const createTextInstance = (content) => {
      return document.createTextNode(content);
  };
  const appendChildToContainer = appendInitalChild;

  let nextEffect = null;
  const commitMutationEffects = (finishedWork) => {
      nextEffect = finishedWork;
      while (nextEffect !== null) {
          const child = nextEffect.child;
          if ((nextEffect.subtreeFlags & MuationMask) !== NoFlags && child !== null) {
              nextEffect = child;
          }
          else {
              up: while (nextEffect !== null) {
                  commitMutationEffectsOnFiber(nextEffect);
                  const sibling = nextEffect.sibling;
                  if (sibling !== null) {
                      nextEffect = sibling;
                      break up;
                  }
                  nextEffect = nextEffect.return;
              }
          }
      }
  };
  const commitMutationEffectsOnFiber = (finishedWork) => {
      const flags = finishedWork.flags;
      if ((flags & Placement) !== NoFlags) {
          commitPlacment(finishedWork);
          finishedWork.flags &= ~Placement;
      }
  };
  const commitPlacment = (finishedWork) => {
      // eslint-disable-next-line no-undef
      {
          console.log("执行Placment操作");
      }
      const hostParent = getHostParaent(finishedWork);
      if (hostParent !== null) {
          appendPlacementNodeIntoContainer(finishedWork, hostParent);
      }
  };
  function getHostParaent(fiber) {
      let parent = fiber.return;
      while (parent) {
          const parentTag = parent.tag;
          if (parentTag === HostComponent) {
              return parent.stateNode;
          }
          if (parentTag === HostRoot) {
              return parent.stateNode.container;
          }
          parent = parent.return;
      }
      // eslint-disable-next-line no-undef
      {
          console.log("未找到");
      }
      return null;
  }
  function appendPlacementNodeIntoContainer(finishedWork, hostParent) {
      if (finishedWork.tag === HostComponent || finishedWork.tag === HostText) {
          appendChildToContainer(finishedWork.stateNode, hostParent);
          return;
      }
      const child = finishedWork.child;
      if (child !== null) {
          appendPlacementNodeIntoContainer(child, hostParent);
          let sibling = child.sibling;
          while (sibling !== null) {
              appendPlacementNodeIntoContainer(sibling, hostParent);
              sibling = sibling.sibling;
          }
      }
  }

  //
  const completeWork = (wip) => {
      const newProps = wip.pendingProps;
      const current = wip.alternate;
      switch (wip.tag) {
          case HostComponent:
              if (current !== null && wip.stateNode) {
                  console.log("");
              }
              else {
                  // const instance = createInstance(wip.type, newProps);
                  const instance = createInstance(wip.type);
                  appendAllChildren(instance, wip);
                  wip.stateNode = wip;
              }
              bubbleProperties(wip);
              return null;
          case HostText:
              if (current !== null && wip.stateNode) {
                  console.log("");
              }
              else {
                  const instance = createTextInstance(newProps);
                  appendAllChildren(instance, wip);
                  wip.stateNode = wip;
              }
              bubbleProperties(wip);
              return null;
          case HostRoot:
              bubbleProperties(wip);
              return null;
          default:
              // eslint-disable-next-line no-undef
              {
                  console.warn("未处理的completeWork", wip);
              }
              break;
      }
  };
  // eslint-disable-next-line no-undef
  function appendAllChildren(parent, wip) {
      let node = wip.child;
      while (node !== null) {
          if (node.tag === HostComponent || node.tag === HostText) {
              appendInitalChild(parent, node.stateNode);
          }
          else if (node.child !== null) {
              node.child.return = node;
              node = node.child;
              continue;
          }
          if (node === wip) {
              return;
          }
          while (node.sibling === null) {
              if (node.return === null && node.return === wip) {
                  return;
              }
              node = node?.return;
          }
          node.sibling.return = node.return;
          node = node.sibling;
      }
  }
  function bubbleProperties(wip) {
      let subtreeFlags = NoFlags;
      let child = wip.child;
      while (child !== null) {
          subtreeFlags |= child.subtreeFlags;
          subtreeFlags |= child.flags;
          child.return = wip;
          child = child.sibling;
      }
      wip.subtreeFlags |= subtreeFlags;
  }

  let workInProgress = null;
  function prepareFreshStack(root) {
      workInProgress = createWorkInProgress(root.current, {});
  }
  function scheduleUpdateOnFilber(fiber) {
      // TODO:调度功能
      const root = markUpdateFromFiberToRoot(fiber);
      renderRoot(root);
  }
  function markUpdateFromFiberToRoot(fiber) {
      let node = fiber;
      let parent = node.return;
      while (parent !== null) {
          node = parent;
          parent = node.return;
      }
      if (node.tag === HostRoot) {
          return node.stateNode;
      }
      return null;
  }
  function renderRoot(root) {
      // 初始化
      prepareFreshStack(root);
      do {
          try {
              workLoop();
              break;
          }
          catch (err) {
              // eslint-disable-next-line no-undef
              {
                  console.warn("workLoop发生错误", err);
              }
              workInProgress = null;
          }
          // eslint-disable-next-line no-constant-condition
      } while (true);
      const finishWork = root.current.alternate;
      root.finishedWork = finishWork;
      commitRoot(root);
  }
  function commitRoot(root) {
      const finishedWork = root.finishedWork;
      if (!finishedWork) {
          return;
      }
      // eslint-disable-next-line no-undef
      {
          console.log("commit阶段开始");
      }
      root.finishedWork = null;
      // 判断是否存在3个子阶段需要执行的操作
      const subtreeHasEffect = (finishedWork.subtreeFlags & MuationMask) !== NoFlags;
      const rooHasEffect = (finishedWork.flags & MuationMask) !== NoFlags;
      if (subtreeHasEffect || rooHasEffect) {
          // beformMutation
          // mutation placement
          commitMutationEffects(finishedWork);
          root.current = finishedWork;
          // layout
      }
      else {
          root.current = finishedWork;
      }
  }
  function workLoop() {
      while (workInProgress !== null) {
          performUnitOfWork(workInProgress);
      }
  }
  function performUnitOfWork(fiber) {
      const next = beginWork(fiber);
      fiber.memoizeProps = fiber.pendingProps;
      if (next === null) {
          completeUnitOfWork(fiber);
      }
      else {
          workInProgress = next;
      }
  }
  function completeUnitOfWork(filber) {
      let node = filber;
      do {
          completeWork(node);
          const sibling = node.sibling;
          if (sibling !== null) {
              workInProgress = sibling;
              return;
          }
          node = node.return;
      } while (node !== null);
  }

  function createContainer(container) {
      const hostRootFilber = new FiberNode(HostRoot, {}, null);
      const root = new FiberRootNode(container, hostRootFilber);
      hostRootFilber.updateQueue = createUpdateQueue();
      return root;
  }
  function updateContainer(element, root) {
      const hostRootFiber = root.current;
      const update = createUpdate(element);
      enqueueUpdate(hostRootFiber.updateQueue, update);
      scheduleUpdateOnFilber(hostRootFiber);
      return element;
  }

  function createRoot(container) {
      const root = createContainer(container);
      return {
          render(element) {
              updateContainer(element, root);
          },
      };
  }

  var ReactDom = /*#__PURE__*/Object.freeze({
    __proto__: null,
    createRoot: createRoot
  });

  return ReactDom;

}));
